[Support][YamlTraits] Add quoting for keys in textual YAML representation

jasilvanus opened the pull request:
The support library contains helpers to parse and emit YAML documents.

In the textual YAML representation, some strings need to be quoted, e.g.
when containing unprintable characters. There are also cases where quoting
isn't strictly necessary, but prevents issues in some cases, for instance strings
consisting of digits only.

We already have such quoting implemented for YAML values.

This patch applies the same quoting to YAML *keys*.

One affected case is output of control registers in AMDGPU Msgpack metadata,
which are printed in a format like this:

```
   0x2cca (SPI_SHADER_PGM_RSRC1_ES): 42
```

With this patch, the key is quoted:

```
   '0x2cca (SPI_SHADER_PGM_RSRC1_ES)': 42
```

Most test changes come from this pattern.

jasilvanus committed:
[Support][YamlTraits] Factor out string quoting [nfc]

Factor out string quoting from Output::scalarString to
a new output(StringRef, QuotingType) helper.

This prepares quoting keys.

jasilvanus committed:
[Support][YamlTraits] Quote keys in textual YAML representation if necessary

The support library contains helpers to parse and emit YAML documents.

In the textual YAML representation, some strings need to be quoted, e.g.
when containing unprintable characters. There are also cases where quoting
isn't strictly necessary, but prevents issues in some cases, for instance strings
consisting of digits only.

We already have such quoting implemented for YAML values.

This patch applies the same quoting to YAML *keys*.

One affected case is output of control registers in AMDGPU Msgpack metadata,
which are printed in a format like this:

   0x2cca (SPI_SHADER_PGM_RSRC1_ES): 42

With this patch, the key is quoted:

   '0x2cca (SPI_SHADER_PGM_RSRC1_ES)': 42

Most test changes come from this pattern.

arsenm reviewed (commented):
Unit tests in llvm/unittests/Support/YAML*?

jasilvanus committed:
[Support][YamlTraits] Add unit tests for quoted keys

jasilvanus committed:
[Support][YamlTraits] Update WebAssembly test

jasilvanus commented:
> Unit tests in llvm/unittests/Support/YAML*?

Good point, thanks. I've added unit tests.

arsenm commented:
> One affected case is output of control registers in AMDGPU Msgpack metadata, which are printed in a format like this:
> 
> ```
>    0x2cca (SPI_SHADER_PGM_RSRC1_ES): 42
> ```
> 
> With this patch, the key is quoted:
> 
> ```
>    '0x2cca (SPI_SHADER_PGM_RSRC1_ES)': 42
> ```

Could we just drop the space in the printing here?

arsenm commented on llvm/lib/Support/YAMLTraits.cpp:
```diff
@@ -801,6 +769,45 @@ void Output::output(StringRef s) {
+    return;
+  }
+
+  const char *const Quote = MustQuote == QuotingType::Single ? "'" : "\"";
```
StringLiteral?

arsenm commented on llvm/lib/Support/YAMLTraits.cpp:
```diff
@@ -801,6 +769,45 @@ void Output::output(StringRef s) {
+    if (S[j] == '\'') {                   // Escape quotes.
+      output(StringRef(&Base[i], j - i)); // "flush".
+      output(StringLiteral("''"));        // Print it as ''
+      i = j + 1;
```
Can you use a StringRef with take_back incrementing? (I guess this is just moving code and shouldn't change here)

jasilvanus replied:
I'd prefer to keep it as-is as this code is just moved.

arsenm commented on llvm/test/Transforms/WholeProgramDevirt/import-indir.ll:
```diff
@@ -3,7 +3,7 @@
 
 ; SUMMARY:     GlobalValueMap:
-; SUMMARY-NEXT:  42:
+; SUMMARY-NEXT:  '42':
```
Should quotes only be emitted in the necessary cases?

arsenm reviewed (commented):
> here are also cases where quoting
isn't strictly necessary, but prevents issues in some cases, for instance strings
consisting of digits only.

What is the issue with just digit only keys?

jasilvanus committed:
Review change: use std::string::npos, use StringLiteral

jasilvanus committed:
[Support][YamlTraits] Don't quote null/numeric/bool keys

For string *values*, we quote null/numeric/bool values
to differentiate from the non-string values.

However, for *keys* this is not necessary, as we only support string
keys for now in the interface, so there is no way to set non-string keys.

Extend the needsQuotes() helper with a new flag that configures whether
quoting to prevent the string type is necessary.

jasilvanus commented:
> What is the issue with just digit only keys?

I re-used the existing `needsQuotes` function that is already used for string values implemented here which quotes numeric strings:

https://github.com/llvm/llvm-project/blob/a02019960b1a693320cd43b0ed6653d95877b94f/llvm/include/llvm/Support/YAMLTraits.h#L674

I originally thought there was some potential issue with numeric keys, but now I'm quite confident this just
to preserve the string type, in order to be able to differentiate the string value `'42'` from the numeric value `42`.

For string values, this totally makes sense as we have typed value setters, so if one wants the unquoted variant one should not set a string value to begin with.
However, keys currently always are strings, and thus there is no way to differentiate the keys `'42'` and `42`.

In the last commit, I extended `needsQuotes` to add a flag whether preservation of string values is required, and not require quoting in the `isNull`/`isBool`/`isNumeric` cases then.

jasilvanus committed:
Merge branch 'main' into jsilvanu/yaml-quoting

jasilvanus commented:
I couldn't repro the test failure locally, trying to merge `main` again and re-trigger CI.

jasilvanus committed:
Merge branch 'main' into jsilvanu/yaml-quoting
